### ISSUE TITLE
Bundles should be produceable as shared objects

### ISSUE THREAD

**Issue by wyessen** [pytorch/glow#3004]
Currently, there is no way to emit bundles as shared libraries. This would be extremely beneficial - this would add the ability to load bundles as "plugins" at runtime on demand (e.g. with dlopen() on UNIX systems). It seems this is currently impossible. Perhaps we could use an additional option, say `-shared`, along with `-emit-bundle` to indicate that we are asking for a shared object.

**Comment by wyessen** [comment]
I have confirmed that it is possible to load bundles as dynamic libraries at runtime. To achieve this, when emitting a bundle, use the `-llvm-compiler=[your compiler]` option together with `-llvm-compiler-opt=-shared`. This will spit out `[your bundle].o`. I was then able to load it with `dlopen()`, and look all the needed symbols in it with `dlsym()`. 

My system is Ubuntu 18.04. A similar setup would probably work on any UNIX and Linux system (including OS X). 

This feels a little like a hack. So the request is really to support an additional `-llvm-shared-lib` option, or something similar, to avoid having to pass the two extra options `-llvm-compiler` and `-llvm-compiler-opt`. Moreover, it isn't obvious out of the box that those two options would do the job (I had to dig around a bit), while having an explicit option like `-llvm-shared-lib` would indicate to clients that bundles _can_ be emitted in the form of dynamic objects, and this functionality is officially supported.

**Comment by opti-mix** [comment]
@wyessen Cool! Very nice finding! I have not thought about producing dynamic libraries this way even though I introduced these options ;-) But of course, it makes sense.

> option like -llvm-shared-lib would indicate to clients that bundles can be emitted in the form of dynamic objects, and this functionality is officially supported

I agree. Having such an option would be much simpler. And it should be pretty straight forward to add this I think.

**Comment by opti-mix** [comment]
@wyessen I discussed this topic with @compnerd who has a lot of experience with linkers, libraries and object file formats. The interesting thing is that Clang/LLVM itself does not produce shared objects directly. Instead they produce an object file and then pass it down to the linker with some special options instructing the linker to create a shared library from this file. The logic for invoking the linker directly is non-trivial and pretty platform-dependent. Adding it to Glow would overcomplicate things. But Clang/LLVM compiler drivers can handle it already. Thus, it seems like invoking the external clang/llvm is the best option to solve the issue. Even if we'd go for invoking the linker directly it would still need two command-line options: one to specify which linker/cross-linker to use and another one to tell Glow that it should produce a shared library instead of an object file. And this seems not any better than the solution you described.

**Comment by wyessen** [comment]
@opti-mix Understood, thanks for the explanation! I guess at this point the best solution would be to document this somewhere with a minimal example.

**Comment by opti-mix** [comment]
@wyessen 

> I guess at this point the best solution would be to document this somewhere with a minimal example.

Totally agree! 

Would you be willing to submit a small PR with such a documentation improvement/example? AOT.md would be a good place to document this.

**Comment by wyessen** [comment]
@opti-mix Yep. I still owe you docs for the other issue (building for special targets). Just need to find the time. I'll get to it...sometime soon hopefully.
